### PR TITLE
Bug/ccp 230

### PR DESCRIPTION
**Pull request by susmitha8**
Normalized the search values

### CODE REVIEW

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 1]
```diff
@@ -636,6 +670,31 @@ private void buildTokenizedOrWildCardQuery(LuceneField luceneFieldAnnotation, St
             }
         }
 
+    /*public class CustomAnalyzer extends Analyzer {
```
Comment (abhikrish503501) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Please remove these comments

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 2]
```diff
@@ -648,6 +707,14 @@ private void buildTokenizedOrWildCardQuery(LuceneField luceneFieldAnnotation, St
     private void createWildCardSearchQuery(LuceneField luceneFieldAnnotation, String searchString,
             BooleanQuery.Builder queryBuilder, Occur occur) {
         if (luceneFieldAnnotation.isWildCardSearchEnabled()) {
+            /*StringBuilder sb = new StringBuilder(searchString.length());
```
Comment (abhikrish503501) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Please remove this comments too

Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Ya, have done that and committed already.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 3]
```diff
@@ -63,6 +76,8 @@
 import org.springframework.core.env.Environment;
 import org.springframework.stereotype.Component;
 
+import com.db4o.query.Query;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Pretty sure that's the wrong query class

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 4]
```diff
@@ -132,7 +148,11 @@
     private Directory index;
     private Path relativePath = null;
     private IndexSearcher searcher = null;
-
+    private Analyzer customAnalyzer() throws IOException{
+        return CustomAnalyzer.builder().withTokenizer("standard").addTokenFilter("asciifolding").
+                addTokenFilter("lowercase").build();
+    }
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: is there a reason that this is a method rather than just a class variable?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 5]
```diff
@@ -497,7 +518,7 @@ public void indexConcepts(String userName)
         if (operator == null || operator.equalsIgnoreCase(SearchParamters.OP_AND)) {
             occur = BooleanClause.Occur.MUST;
         }
-
+        try {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: where is the catch block for this try? Give that I can't see it, it probably means the try block is waayyyy to big. Catch exceptions where they occur and deal with them appropriately right where they are thrown.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 6]
```diff
@@ -530,14 +551,23 @@ public void indexConcepts(String userName)
...
-                    buildQuery(occur, perFieldAnalyzerWrapper, qBuild, builder, luceneFieldAnnotation, searchString);
+                    TokenStream normalizedValue = customAnalyzer().tokenStream("fieldName", searchString);
+                    CharTermAttribute cattr = normalizedValue.addAttribute(CharTermAttribute.class);
+                    normalizedValue.reset();
+                    String normalizedSearchString = null;
+                    while (normalizedValue.incrementToken()) {
+                        normalizedSearchString = cattr.toString();
+                    }
+                    normalizedValue.end();
+                    normalizedValue.close();
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: if you set the customanalyzer into the perFieldAnalyzerWrapper then you shouldn't need this. the query will take care of it.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 7]
```diff
@@ -579,13 +609,14 @@ public void indexConcepts(String userName)
                 ConceptEntry entry = getConceptFromDocument(d);
                 concepts.add(entry);
             }
+            logger.debug("Number of concepts retrieved from lucene = " + concepts.size());
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: remove

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 8]
```diff
@@ -622,15 +653,20 @@ private void buildQuery(BooleanClause.Occur occur, PerFieldAnalyzerWrapper perFi
...
+        hypSearchStr = searchString.replace('-', ' ');
             if (luceneFieldAnnotation.isWildCardSearchEnabled()) {
+                for (String searchValue : hypSearchStr.split(" ")) {
                 createWildCardSearchQuery(luceneFieldAnnotation, searchValue, tokenizedQueryBuilder, Occur.MUST);
+                }
             } else {
+                for (String searchValue : hypSearchStr.split(" ")) {
                 tokenizedQueryBuilder.add(new PhraseQuery(luceneFieldAnnotation.lucenefieldName(), searchValue),
                         Occur.MUST);
+                }
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: why is this still here? this should not be needed with the custom anlayzer.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ConceptSearchController.java [hunk 9]
```diff
@@ -87,6 +87,8 @@ public String search(HttpServletRequest req, ModelMap model, @RequestParam(defau
         if (results.hasErrors()) {
             return "conceptsearch";
         }
+        String str = conceptSearchBean.getWord().trim();
+        conceptSearchBean.setWord(str);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ConceptSearchController.java]: isnt' that a different ticket? why is that here?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 8 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: we talked about not instantiating query objects, but use the builder to be able to add the analyzers per field. Why did you do that? Was there a problem?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 3 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Yea. Never used that db4 query class, imported by mistake and it was never used. Have used org.apache.lucene.search.Query. Will remove the unused import.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 4 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: when declaring it as the class variable its throwing the error "While constructing the default constructor can not handle exception : type Exception thrown by implicit super constructor". To handle the exception I used method instead of a class variable.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 5 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Yes, it is because the try block is big. Will handle this.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 6 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Yea, this normalizer is no longer required and doesn't serve the purpose in case of exact search. Will remove this.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 4 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Initialize a class variable analyzer in an init method and you should be fine.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 7 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: isn't this logger required to log the number of concepts retrieved for the query? This was there from before.

Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: ya, that was never a really useful log message...

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 8 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Yes, we discussed about instantiating the objects using builder so that analyzer can be applied to the fields and I am working on it. This pull request was raised before our discussion, so it wasn't implemented.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ConceptSearchController.java (hunk 9 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ConceptSearchController.java]: This was a mistake. Instead of creating a branch out of develop, I think I created a branch out of CCP 243 and that is why that change is here and as I addressed one of the review comments and committed the code its here. Sorry about this.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 7 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Will remove the log message.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 10]
```diff
@@ -517,10 +536,10 @@ public void indexConcepts(String userName)
 
         }
 
+        
         PerFieldAnalyzerWrapper perFieldAnalyzerWrapper = new PerFieldAnalyzerWrapper(standardAnalyzer,
-                analyzerPerField);
-
-        QueryBuilder qBuild = new QueryBuilder(perFieldAnalyzerWrapper);
+                    analyzerPerField);
+        QueryBuilder qBuild = new QueryBuilder(customAnalyzer);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: this should still have the perfield analyzer

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 11]
```diff
@@ -530,7 +549,7 @@ public void indexConcepts(String userName)
                 String searchString = fieldMap.get(search.fieldName());
                 if (searchString != null) {
                     searchString = searchString.toLowerCase();
-                    buildQuery(occur, perFieldAnalyzerWrapper, qBuild, builder, luceneFieldAnnotation, searchString);
+                    buildQuery(occur, perFieldAnalyzerWrapper, qBuild, query, builder, luceneFieldAnnotation, searchString);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: what is 'query' used for? why is it a class variable? this is a multi-threaded environment, having a class variable that is changed in a method is a really bad idea and will cause all kinds of issues.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 12]
```diff
@@ -132,19 +146,24 @@
     private Directory index;
     private Path relativePath = null;
     private IndexSearcher searcher = null;
-
+    private Analyzer customAnalyzer = null;
+    private Query query = null;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: what's that used for?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 13]
```diff
@@ -613,24 +631,26 @@ private void buildQuery(BooleanClause.Occur occur, PerFieldAnalyzerWrapper perFi
...
     }
 
-    private void buildTokenizedOrWildCardQuery(LuceneField luceneFieldAnnotation, String searchString,
-            BooleanQuery.Builder tokenizedQueryBuilder) {
-        for (String searchValue : searchString.split(" ")) {
+    private void buildTokenizedOrWildCardQuery(LuceneField luceneFieldAnnotation, String searchString, QueryBuilder qBuild,
+            BooleanQuery.Builder tokenizedQueryBuilder) {       
             if (luceneFieldAnnotation.isWildCardSearchEnabled()) {
-                createWildCardSearchQuery(luceneFieldAnnotation, searchValue, tokenizedQueryBuilder, Occur.MUST);
+                BooleanQuery.Builder analyzedbuilder = new BooleanQuery.Builder();
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: variable names use camel case

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ConceptSearchController.java (hunk 9 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ConceptSearchController.java]: please remove and commit to the correct branch

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 13 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: changed the variable name

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 12 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: remove the query variable which isn't being used anywhere. The customAnalyzer is used by add ascii folding filter to it. The variable is declared there

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 11 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: removed the query variable

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 10 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: removed customAnalyzer and added the perfieldanalyzer

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ConceptSearchController.java (hunk 9 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/web/ConceptSearchController.java]: removed the change from this branch

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 14]
```diff
@@ -613,24 +628,26 @@ private void buildQuery(BooleanClause.Occur occur, PerFieldAnalyzerWrapper perFi
         } else {
             if (luceneFieldAnnotation.isWildCardSearchEnabled()) {
                 createWildCardSearchQuery(luceneFieldAnnotation, searchString, builder, occur);
-            } else {
-                builder.add(new BooleanClause(
-                        new TermQuery(new Term(luceneFieldAnnotation.lucenefieldName(), searchString)), occur));
             }
+                builder.add(new BooleanClause(
```
[the remark this answers is not in the document]
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: indented

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 10 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: Yup, Got it. replaced the standardAnalyzer.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java [hunk 15]
```diff
@@ -579,58 +595,55 @@ public void indexConcepts(String userName)
...
-
+                shortWordSearchQueryBuilder.add(new PhraseQuery(luceneFieldAnnotation.lucenefieldName() + LuceneFieldNames.UNTOKENIZED_SUFFIX, searchString), Occur.SHOULD);
                 rootQueryBuilder.add(shortWordSearchQueryBuilder.build(), Occur.SHOULD);
                 tokenizedQueryBuilder = rootQueryBuilder;
             }
             builder.add(tokenizedQueryBuilder.build(), occur);
 
-        } else {
+        } 
+        else {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: please keep our code style

**Comment by jdamerow** [comment]
ok to test

**Comment by jdamerow** [comment]
awesome, works great! There is a little issue with wildcards and umlauts, but that can be addressed in a later ticket. for now just fix the formatting, and we should be good to go.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java (hunk 15 above)
Comment (susmitha8) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/lucene/impl/LuceneUtility.java]: changed
